### ISSUE TITLE
ERROR GC overhead limit exceeded

### ISSUE THREAD

**Issue by elyisgreat** [alda-lang/alda#257]
I found a very strange bug with one of the pieces I've written

```
(tempo! 100)

bPart1 = o5 r1 (quant 30) c8 f f c d~ c16~ <b-~> c8 < a b-~ a16~ g~ a8 f e g > c4~
bPart2 = o2 (quant 30) [f8 >c<]*10 [c g]*2

part1 = o5 (quant 30) c8 f f c d~ c16~ <b-~> c8 < a b-~ a16~ g~ a8 f e g32~ a~ g~ a~ f8~ c16~ d
part2 = o2 (quant 30) [f8 >c<]*6 [c g]*2

part1 = part1 [e8~ f16~ g~ e8 c]*3 >c8~< b-16~ a~ g~ a~ g~ f~ [e8~ f16~ g~ e8 c]*3 >c8.~< b-16~ a~ g~ <a~>/a~ <b-~>/b-~ >
part2 = part2 (quant 60) [[g2/>c/e<]*2 [g4/>c/e<]*2 g2/>c/e<]*2 (quant 30)

part1 = part1 <c8>/c <f>/f <f>/f <c>/c <d~>/d~ <c16~>/c~ < <b-~>/b-~ > <c8>/c < <a>/a <b-~>/b-~ <a>/a16~ <g~>/g~ <a8>/a <f>/f <e>/e <g>/g > <c4~>/c~
part2 = part2 [f8 a/>c<]*6 [c g/>c<]*2

part1 = part1 <c8>/c <f>/f <f>/f <c>/c <d~>/d~ <c16~>/c~ < <b-~>/b-~ > <c8>/c < <a>/a <b-~>/b-~ <a>/a16~ <g~>/g~ <a8>/a <f>/f <e>/e <g>/g <f~>/f~ (quant 90) e16 f
part2 = part2 [f8 a/>c<]*6 [c g/>c<]*2

part1 = part1 g8 > c c < g a g16 f g8 e >> c < b-16 a b-8 g a b- > c d- < c f f c d c16 < b- > c8 < a > b a16 g a8 f d2/<g8 f16 e f8 d r4. d16 e f f+ g a g24/b- a/>c< b/>d c/e d/f e-/g-
part2 = part2 c1~/g~/>c~< e-~/g~/b-~/>e-~< f~/a~/>c~/f~< g~/b~/>d~/g~ (vol 90) d8~ c16~ < b~ > c8~ < a~ r2 (vol 60)

part1 = part1 <g8>/e/g <f16>/d/f <e>/c/e <f8>/d/f <d/g/>d <e>/c/e <c/g>/c < <g>/e/g > <c/g>/c <d/b>/d <e16>/c/e <f>/d/f <e8>/c/e <d/b>/d <c/g>/c < <a>/e/a <g4>/e/g >
part2 = part2 [c8 g/>e/g<]*4 [g >c/f<]*2 [c8 g/>e/g<]*2

part1 = part1 <g8>/e/g <f16>/d/f <e>/c/e <f8>/d/f <d/g/>d <e>/c/e <c/g>/c < <g>/e/g > <c/g>/c <d/b>/d <f>/d/f <e32/b>/e <f>/d/f <e16/b>/e <d8/f>/d <c2/e>/c
part2 = part2 [c8 g/>e/g<]*4 [g >c/f<]*2 [c8 g/>e/g<]*2

part1 = part1 c8/>c< <b-16>/b- <a>/a <b-8>/b- <g>/g <a>/a <f>/f <c>/c <f>/f <g>/g <a16>/a <b->/b- <a8>/a <g>/g <f>/f <d>/d <c4>/c
part2 = part2 [f8 a/>c<]*4 [c g/>c<]*2 [f a/>c<]*2

part1 = part1 c8/>c< <b-16>/b- <a>/a <b-8>/b- <g>/g <a>/a <f>/f <c>/c <f>/f <g>/g <b->/b- <a>/a <g>/g <f2>/f
part2 = part2 [f8 a/>c<]*4 [c g/>c<]*2 [f a/>c<]*2

part1 = part1 (quant 30) c8 f f c d~ c16~ <b-~> c8 < a b-~ a16~ g~ a8 f e g8/e16~ f~ > c4~/<f+16~ g~ a~ b-~
part2 = part2 [f8 >c<]*6 [c g]*2

ePart1 = o5 (quant 30) c8 f f c d~ c16~ <b-~> c8 < a/>a< b-~/>b-~< a16~/>a~< g~/>g~< a8/>a< f/>f< (tempo 72.5) e/>e< g/>g< f4~1~/>f~<
ePart2 = o2 (quant 30) [f8 >c<]*6 (tempo 72.5) c g f4~1~/>c~/f~<

piano:

V1: (vol 90) bPart1 [part1]*1 ePart1
V2: (vol 60) bPart2 [part2]*1 ePart2
```

For some reason, the score loads indefinitely. However, just by commenting 2 lines out

```
(tempo! 100)

bPart1 = o5 r1 (quant 30) c8 f f c d~ c16~ <b-~> c8 < a b-~ a16~ g~ a8 f e g > c4~
bPart2 = o2 (quant 30) [f8 >c<]*10 [c g]*2

part1 = o5 (quant 30) c8 f f c d~ c16~ <b-~> c8 < a b-~ a16~ g~ a8 f e g32~ a~ g~ a~ f8~ c16~ d
part2 = o2 (quant 30) [f8 >c<]*6 [c g]*2

part1 = part1 [e8~ f16~ g~ e8 c]*3 >c8~< b-16~ a~ g~ a~ g~ f~ [e8~ f16~ g~ e8 c]*3 >c8.~< b-16~ a~ g~ <a~>/a~ <b-~>/b-~ >
part2 = part2 (quant 60) [[g2/>c/e<]*2 [g4/>c/e<]*2 g2/>c/e<]*2 (quant 30)

part1 = part1 <c8>/c <f>/f <f>/f <c>/c <d~>/d~ <c16~>/c~ < <b-~>/b-~ > <c8>/c < <a>/a <b-~>/b-~ <a>/a16~ <g~>/g~ <a8>/a <f>/f <e>/e <g>/g > <c4~>/c~
part2 = part2 [f8 a/>c<]*6 [c g/>c<]*2

part1 = part1 <c8>/c <f>/f <f>/f <c>/c <d~>/d~ <c16~>/c~ < <b-~>/b-~ > <c8>/c < <a>/a <b-~>/b-~ <a>/a16~ <g~>/g~ <a8>/a <f>/f <e>/e <g>/g <f~>/f~ (quant 90) e16 f
part2 = part2 [f8 a/>c<]*6 [c g/>c<]*2

#part1 = part1 g8 > c c < g a g16 f g8 e >> c < b-16 a b-8 g a b- > c d- < c f f c d c16 < b- > c8 < a > b a16 g a8 f d2/<g8 f16 e f8 d r4. d16 e f f+ g a g24/b- a/>c< b/>d c/e d/f e-/g-
#part2 = part2 c1~/g~/>c~< e-~/g~/b-~/>e-~< f~/a~/>c~/f~< g~/b~/>d~/g~ (vol 90) d8~ c16~ < b~ > c8~ < a~ r2 (vol 60)

part1 = part1 <g8>/e/g <f16>/d/f <e>/c/e <f8>/d/f <d/g/>d <e>/c/e <c/g>/c < <g>/e/g > <c/g>/c <d/b>/d <e16>/c/e <f>/d/f <e8>/c/e <d/b>/d <c/g>/c < <a>/e/a <g4>/e/g >
part2 = part2 [c8 g/>e/g<]*4 [g >c/f<]*2 [c8 g/>e/g<]*2

part1 = part1 <g8>/e/g <f16>/d/f <e>/c/e <f8>/d/f <d/g/>d <e>/c/e <c/g>/c < <g>/e/g > <c/g>/c <d/b>/d <f>/d/f <e32/b>/e <f>/d/f <e16/b>/e <d8/f>/d <c2/e>/c
part2 = part2 [c8 g/>e/g<]*4 [g >c/f<]*2 [c8 g/>e/g<]*2

part1 = part1 c8/>c< <b-16>/b- <a>/a <b-8>/b- <g>/g <a>/a <f>/f <c>/c <f>/f <g>/g <a16>/a <b->/b- <a8>/a <g>/g <f>/f <d>/d <c4>/c
part2 = part2 [f8 a/>c<]*4 [c g/>c<]*2 [f a/>c<]*2

part1 = part1 c8/>c< <b-16>/b- <a>/a <b-8>/b- <g>/g <a>/a <f>/f <c>/c <f>/f <g>/g <b->/b- <a>/a <g>/g <f2>/f
part2 = part2 [f8 a/>c<]*4 [c g/>c<]*2 [f a/>c<]*2

#part1 = part1 (quant 30) c8 f f c d~ c16~ <b-~> c8 < a b-~ a16~ g~ a8 f e g8/e16~ f~ > c4~/<f+16~ g~ a~ b-~
#part2 = part2 [f8 >c<]*6 [c g]*2

ePart1 = o5 (quant 30) c8 f f c d~ c16~ <b-~> c8 < a/>a< b-~/>b-~< a16~/>a~< g~/>g~< a8/>a< f/>f< (tempo 72.5) e/>e< g/>g< f4~1~/>f~<
ePart2 = o2 (quant 30) [f8 >c<]*6 (tempo 72.5) c g f4~1~/>c~/f~<

piano:

V1: (vol 90) bPart1 [part1]*1 ePart1
V2: (vol 60) bPart2 [part2]*1 ePart2
```

It does not load indefinitely. 

It doesn't seem to matter which pair of part1/part2 variable declarations I comment out, however including all of them will make it unable to play. 

```
(tempo! 100)

bPart1 = o5 r1 (quant 30) c8 f f c d~ c16~ <b-~> c8 < a b-~ a16~ g~ a8 f e g > c4~
bPart2 = o2 (quant 30) [f8 >c<]*10 [c g]*2

part1 = o5 (quant 30) c8 f f c d~ c16~ <b-~> c8 < a b-~ a16~ g~ a8 f e g32~ a~ g~ a~ f8~ c16~ d
part2 = o2 (quant 30) [f8 >c<]*6 [c g]*2

part1 = part1 [e8~ f16~ g~ e8 c]*3 >c8~< b-16~ a~ g~ a~ g~ f~ [e8~ f16~ g~ e8 c]*3 >c8.~< b-16~ a~ g~ <a~>/a~ <b-~>/b-~ >
part2 = part2 (quant 60) [[g2/>c/e<]*2 [g4/>c/e<]*2 g2/>c/e<]*2 (quant 30)

part1 = part1 <c8>/c <f>/f <f>/f <c>/c <d~>/d~ <c16~>/c~ < <b-~>/b-~ > <c8>/c < <a>/a <b-~>/b-~ <a>/a16~ <g~>/g~ <a8>/a <f>/f <e>/e <g>/g > <c4~>/c~
part2 = part2 [f8 a/>c<]*6 [c g/>c<]*2

part1 = part1 <c8>/c <f>/f <f>/f <c>/c <d~>/d~ <c16~>/c~ < <b-~>/b-~ > <c8>/c < <a>/a <b-~>/b-~ <a>/a16~ <g~>/g~ <a8>/a <f>/f <e>/e <g>/g <f~>/f~ (quant 90) e16 f
part2 = part2 [f8 a/>c<]*6 [c g/>c<]*2

part1 = part1 g8 > c c < g a g16 f g8 e >> c < b-16 a b-8 g a b- > c d- < c f f c d c16 < b- > c8 < a > b a16 g a8 f d2/<g8 f16 e f8 d r4. d16 e f f+ g a g24/b- a/>c< b/>d c/e d/f e-/g-
part2 = part2 c1~/g~/>c~< e-~/g~/b-~/>e-~< f~/a~/>c~/f~< g~/b~/>d~/g~ (vol 90) d8~ c16~ < b~ > c8~ < a~ r2 (vol 60)

part1 = part1 <g8>/e/g <f16>/d/f <e>/c/e <f8>/d/f <d/g/>d <e>/c/e <c/g>/c < <g>/e/g > <c/g>/c <d/b>/d <e16>/c/e <f>/d/f <e8>/c/e <d/b>/d <c/g>/c < <a>/e/a <g4>/e/g >
part2 = part2 [c8 g/>e/g<]*4 [g >c/f<]*2 [c8 g/>e/g<]*2

part1 = part1 <g8>/e/g <f16>/d/f <e>/c/e <f8>/d/f <d/g/>d <e>/c/e <c/g>/c < <g>/e/g > <c/g>/c <d/b>/d <f>/d/f <e32/b>/e <f>/d/f <e16/b>/e <d8/f>/d <c2/e>/c
part2 = part2 [c8 g/>e/g<]*4 [g >c/f<]*2 [c8 g/>e/g<]*2

part1 = part1 c8/>c< <b-16>/b- <a>/a <b-8>/b- <g>/g <a>/a <f>/f <c>/c <f>/f <g>/g <a16>/a <b->/b- <a8>/a <g>/g <f>/f <d>/d <c4>/c
part2 = part2 [f8 a/>c<]*4 [c g/>c<]*2 [f a/>c<]*2

part1 = part1 c8/>c< <b-16>/b- <a>/a <b-8>/b- <g>/g <a>/a <f>/f <c>/c <f>/f <g>/g <b->/b- <a>/a <g>/g <f2>/f
part2 = part2 [f8 a/>c<]*4 [c g/>c<]*2 [f a/>c<]*2

part1 = part1 (quant 30) c8 f f c d~ c16~ <b-~> c8 < a b-~ a16~ g~ a8 f e g8/e16~ f~ > c4~/<f+16~ g~ a~ b-~
part2 = part2 [f8 >c<]*6 [c g]*2

ePart1 = o5 (quant 30) c8 f f c d~ c16~ <b-~> c8 < a/>a< b-~/>b-~< a16~/>a~< g~/>g~< a8/>a< f/>f< (tempo 72.5) e/>e< g/>g< f4~1~/>f~<
ePart2 = o2 (quant 30) [f8 >c<]*6 (tempo 72.5) c g f4~1~/>c~/f~<

piano:

V1: (vol 90) bPart1 [part1]*1 ePart1
V2: (vol 60) bPart2 [part2]*1 ePart2
```

At first I thought it was an issue with having too many variable redeclarations, however I made a test piece with dozens of redeclarations and it ran quite quickly without a hitch:

```
foo = c

foo = foo d
foo = foo e
foo = foo f
foo = foo g
foo = foo a
foo = foo b
foo = foo c
foo = foo d
foo = foo e
foo = foo f
foo = foo g
foo = foo a
foo = foo b
foo = foo c
foo = foo d
foo = foo e
foo = foo f
foo = foo g
foo = foo a
foo = foo b
foo = foo c
foo = foo d
foo = foo e
foo = foo f
foo = foo g
foo = foo a
foo = foo b
foo = foo c
foo = foo d
foo = foo e
foo = foo f
foo = foo g
foo = foo a
foo = foo b
foo = foo c
foo = foo d
foo = foo e
foo = foo f
foo = foo g
foo = foo a
foo = foo b
foo = foo c
foo = foo d
foo = foo e
foo = foo f
foo = foo g
foo = foo a
foo = foo b
foo = foo c
foo = foo d
foo = foo e
foo = foo f
foo = foo g
foo = foo a
foo = foo b
foo = foo c
foo = foo d
foo = foo e
foo = foo f
foo = foo g
foo = foo a
foo = foo b
foo = foo c
foo = foo d
foo = foo e
foo = foo f
foo = foo g
foo = foo a
foo = foo b
foo = foo c
foo = foo d
foo = foo e
foo = foo f
foo = foo g
foo = foo a
foo = foo b
foo = foo c
foo = foo d
foo = foo e
foo = foo f
foo = foo g
foo = foo a
foo = foo b
foo = foo c
foo = foo d
foo = foo e
foo = foo f
foo = foo g
foo = foo a
foo = foo b
foo = foo c
foo = foo d
foo = foo e
foo = foo f
foo = foo g
foo = foo a
foo = foo b
foo = foo c
foo = foo d
foo = foo e
foo = foo f
foo = foo g
foo = foo a
foo = foo b

piano: foo
```

I haven't yet found a base case, and I know I'm not making any syntax errors. What is going on?

EDIT: Loads indefinitely until `[27713] ERROR GC overhead limit exceeded` is thrown.

**Comment by daveyarwood** [comment]
The original score plays fine on my machine, although it takes forever and a day to load... once it did, it at least brought back fond memories of playing Super Mario RPG as a kid :)

I think what we're running into here is that the Java VM is not allocating enough memory for all the score data. Now that we're storing not only events, but also the values of variables, scores are starting to take up a lot more space in memory. Part of the problem is that we're asking Alda to hold the entire score in memory as a big ol' hash-map, when we could be doing things a lot more efficiently.

It's a problem that we would inevitably have to deal with... It's a little bit complex, but it's about time we tackled it.

I think the solution might be, instead of holding all of the data in memory, we could try storing it in a lightweight database like redis or sqlite3. When the score is parsed and evaluated, the events (as well as the values of variables) could be stored as database records, and then when the sound engine is ready to play the score, it could read the events back from the database. I have a feeling this might also help with performance, but we'll see. The immediate problem we're trying to solve is that scores are getting too big to fit in memory.

**Comment by elyisgreat** [comment]
Hm. I wonder why it's causing a GC overhead limit error on my computer... Then again, my computer isn't exactly beefy. I'm curious as to what I'm doing here that I'm not doing in my test score. I also have scores with far more events (but no variables) that load much faster/without throwing this error.

**Comment by daveyarwood** [comment]
I think it's just a matter of where the upper limit is in your JVM. Once it reaches that limit, you end up getting the GC overhead limit error. That might explain why your first score would play if you commented out a couple of lines -- it may have been just over the limit.

I think scores with variables take up more space because not only are the values of each variable stored, but then every time a variable is used, all of its events are added to the score.

We are reaching the point where an Alda score is its own complex little environment involving the states of instrument attributes, potentially huge buckets of events, and most recently, the values of variables, which can themselves be large. This all makes me think that storing the state in a database instead of in a huge, in-memory data structure, is going to be the way to go.

**Comment by elyisgreat** [comment]
This also seems to be very different from #235 , although I'm not sure how.

**Comment by daveyarwood** [comment]
Yeah, the error we encountered in #235 ("Method code too large!") is a weird beast. That had to do with runtime eval of Clojure code -- under the hood, I think Clojure's `eval` creates Java methods and evaluates them, and there is a limit to the size of a Java method, so that error can occur when you try to evaluate a huge chunk of Clojure code.

This is different -- it has to do with the heap size in the JVM. Basically, the JVM sets aside a certain amount of memory that will be available when a program is running -- how much space depends on your system and JVM options. The JVM has garbage collection (GC), which means it deletes stuff whenever it's no longer needed. So most of the time, heap size isn't something we need to worry about, because the GC ensures that we'll always have enough memory for the program to continue doing what it needs to do. But sometimes, if you're dealing with large amounts of data being stored in-memory, you end up running out of heap space.

**Comment by elyisgreat** [comment]
What's especially weird is it seems to be mainly a server issue. if you try loading the song into the REPL, it should load and play without issue.

**Comment by daveyarwood** [comment]
That makes sense -- the server has additional overhead compared to the REPL -- it's running an HTTP server and managing potentially multiple scores at once. So it makes sense if the server runs out of memory more easily.

**Comment by elyisgreat** [comment]
The difference seems to be a lot, though. What the server will load in five minutes and then crash the REPL will load in under a second...

**Comment by daveyarwood** [comment]
Well, that's interesting. Will have to dig in when I have more time.

**Comment by elyisgreat** [comment]
seems to be fixed in rc31. what was changed?

**Comment by daveyarwood** [comment]
OK, I figured it out! There was actually a bug, and it's fixed in [1.0.0-rc31](https://github.com/alda-lang/alda/releases/tag/1.0.0-rc31).

The short explanation: the issue was that the parser 

The long explanation follows... buckle up, this is probably going to be a little confusing.

I think a bug may have been introduced when we were optimizing the parser recently ( #238 ). Basically, the parser is now optimized for parsing entire scores, but is not so good at parsing the components of a score individually, if it's a lot of text and not in the structure that it expects. This is usually not a problem, unless you're trying to get the parser to parse large inputs like an entire score, using the rules for one of those individual components... which is what we were doing by accident, but we never noticed before because the parser used to be just as good at parsing whole scores as individual components.

(By the way, by "individual components," I mean things like the set of events for just one instrument part.)

Before this release, the server would behave like the REPL, in that it would try to determine the "parsing context" (e.g. am I parsing additional notes for an existing instrument part? am I parsing an entire instrument part? am I parsing an entire score?) by starting at the lowest level (notes and other events) and working its way up if the input didn't parse correctly at that level. This works fine for smaller lines of input, like what you would typically enter at the command line (when appending to a working score with the `-a`/`--append` option) or in the REPL. But the problem is, the server was also going through this same routine when you fed it an entire file as a "one-off" score, which is the default behavior when using the command-line client (e.g. `alda play -f my-file.alda`.

So it was trying to parse an entire score as if it were a sequence of events, and getting getting super confused and having to backtrack a lot. Not only can this take a long time, but the parser also has to keep track of all of these passes through the same input, which can cause the JVM to run out of heap space.

The fix in 1.0.0-rc31 is that when you're playing a "one-off" score (which is always the case when using `alda play` without the `--append` option), the parser will now skip straight to parsing the input as if it's an entire score, which is where it would have ended up eventually anyway.
